src/sentry/integrations/vsts_extension/integration.py

```diff
@@ -0,0 +1,51 @@
+from __future__ import absolute_import
+
+from django.contrib import messages
+from django.http import HttpResponseRedirect
+
+from sentry.integrations.vsts.integration import VstsIntegrationProvider
+from sentry.pipeline import PipelineView
+from sentry.utils.http import absolute_uri
+
+
+class VstsExtensionIntegrationProvider(VstsIntegrationProvider):
+    key = 'vsts-extension'
+
+    def get_pipeline_views(self):
+        views = super(VstsExtensionIntegrationProvider, self).get_pipeline_views()
+        views[-1] = VstsExtensionFinishedView()
+        return views
+
+    def build_integration(self, state):
+        # Normally this is saved into the ``identity`` state, but for some
+        # reason it gets wiped out in the NestedPipeline. Instead, we'll store
+        # it in it's own key (``vsts``) to be used down the line in
+        # ``VSTSOrganizationSelectionView``.
+        state['account'] = {
+            'AccountId': state['vsts']['AccountId'],
+            'AccountName': state['vsts']['AccountName'],
+        }
+
+        state['instance'] = '{}.visualstudio.com'.format(
+            state['vsts']['AccountName']
+        )
+
+        return super(
+            VstsExtensionIntegrationProvider,
+            self
+        ).build_integration(state)
+
+
+class VstsExtensionFinishedView(PipelineView):
+    def dispatch(self, request, pipeline):
+        pipeline.finish_pipeline()
```

Comment:
Not sure why you're doing this, but normally:

finish_pipeline isn't meant to be called by views (or called by you at all). You define a view that services the endpoint, and all this view does is fetch the pipeline and call next_step() on it. This call will in turn call your provided views, which should simply fetch the pipeline and do their job. finish_pipeline will be called automatically when all views are exhausted.

It seems like we should change the API so that providers are able to provide their own finish_pipeline definition (since the flows are different, I don't know why the finishing step should be common). e.g. in getsentry/web/identity I make sure finish_pipeline is never even called (it would be wrong to do so), and my views handle everything.

Reply:
This pipeline is a bit different than the others, since the process starts from the VSTS Marketplace and not the Integrations page.

There's no dialog that appears, which means the final request made to Sentry to finish things up (when the dialog closes) is never made. That final request is what programmatically calls `finish_pipeline` normally. Since that doesn't happen, we need to explicitly call it here.